src/core/border_router/routing_manager.cpp

```diff
@@ -0,0 +1,761 @@
+    mAdvertisedOmrPrefixNum = 0;
+
+    mAdvertisedOnLinkPrefix = nullptr;
+
+    mDiscoveredOnLinkPrefix.Clear();
+    mDiscoveredOnLinkPrefixInvalidTimer.Stop();
+
+    mRouterAdvertisementTimer.Stop();
+    mRouterAdvertisementCount = 0;
+
+    mRouterSolicitTimer.Stop();
+    mRouterSolicitCount = 0;
+}
+
+void RoutingManager::RecvIcmp6Message(uint32_t            aInfraIfIndex,
+                                      const Ip6::Address &aSrcAddress,
+                                      const uint8_t *     aBuffer,
+                                      uint16_t            aBufferLength)
+{
+    const Ip6::Icmp::Header *icmp6Header;
+
+    // The Border Routing function is enabled only when we are attached.
+    VerifyOrExit(IsInitialized() && Get<Mle::MleRouter>().IsAttached());
+
+    VerifyOrExit(aInfraIfIndex == mInfraIfIndex);
+    VerifyOrExit(aBuffer != nullptr && aBufferLength >= sizeof(*icmp6Header));
+
+    icmp6Header = reinterpret_cast<const Ip6::Icmp::Header *>(aBuffer);
+
+    switch (icmp6Header->GetType())
+    {
+    case Ip6::Icmp::Header::kTypeRouterAdvert:
+        HandleRouterAdvertisement(aSrcAddress, aBuffer, aBufferLength);
+        break;
+    case Ip6::Icmp::Header::kTypeRouterSolicit:
+        HandleRouterSolicit(aSrcAddress, aBuffer, aBufferLength);
+        break;
+    default:
+        ExitNow();
+    }
+
+exit:
+    return;
+}
+
+void RoutingManager::HandleNotifierEvents(Events aEvents)
+{
+    VerifyOrExit(IsInitialized());
+
+    if (aEvents.Contains(kEventThreadRoleChanged))
+    {
+        if (Get<Mle::MleRouter>().IsAttached())
+        {
+            Start();
+        }
+        else
+        {
+            Stop();
+        }
+    }
+
+    if (aEvents.Contains(kEventThreadNetdataChanged))
+    {
+        if (Get<Mle::MleRouter>().IsAttached())
+        {
+            EvaluateRoutingPolicy();
+        }
+    }
+
+exit:
+    return;
+}
+
+uint8_t RoutingManager::EvaluateOmrPrefix(Ip6::Prefix *aNewOmrPrefixes, uint8_t aMaxOmrPrefixNum)
+{
+    uint8_t                         newOmrPrefixNum = 0;
+    NetworkData::Iterator           iterator        = NetworkData::kIteratorInit;
+    NetworkData::OnMeshPrefixConfig onMeshPrefixConfig;
+    bool                            isLocalOmrPrefixPublished = false;
+
+    VerifyOrExit(Get<Mle::MleRouter>().IsAttached());
+
+    while (Get<NetworkData::Leader>().GetNextOnMeshPrefix(iterator, onMeshPrefixConfig) == OT_ERROR_NONE)
+    {
+        if (!IsValidOmrPrefix(onMeshPrefixConfig.GetPrefix()) || !onMeshPrefixConfig.mDefaultRoute ||
+            !onMeshPrefixConfig.mSlaac || onMeshPrefixConfig.mDp)
+        {
+            continue;
+        }
+
+        if (onMeshPrefixConfig.GetPrefix() == mLocalOmrPrefix)
+        {
+            // We didn't immediately add local OMR prefix here so that
+            // we don't need to remove it later when we want to unpublish it.
+            isLocalOmrPrefixPublished = true;
```

Comment:
We should know if the local OMR prefix should be preferred as the only OMR prefix in Network Data.